lib/ash/actions/read.ex

```diff
@@ -1303,39 +1303,44 @@ defmodule Ash.Actions.Read do
   end
 
   defp run_before_action(query) do
-    query = Ash.Query.put_context(query, :private, %{in_before_action?: true})
+    query =
+      query
+      |> Ash.Query.put_context(:private, %{in_before_action?: true})
+      |> set_phase(:before_action)
 
     query.before_action
     |> Enum.reduce_while({query, []}, fn before_action, {query, notifications} ->
       case before_action.(query) do
         {%{valid?: false} = query, _} ->
-          {:halt, {query, []}}
+          {:halt, {set_phase(query), []}}
 
         %{valid?: false} = query ->
-          {:halt, {query, []}}
+          {:halt, {set_phase(query), []}}
 
         {query, new_notifications} ->
-          {:cont, {query, notifications ++ new_notifications}}
+          {:cont, {set_phase(query), notifications ++ new_notifications}}
 
         query ->
-          {:cont, {query, notifications}}
+          {:cont, {set_phase(query), notifications}}
       end
     end)
   end
 
   defp run_after_action(query, results) do
+    query = set_phase(query, :after_action)
+
     query.after_action
     |> Enum.reduce_while({query, {:ok, results, []}}, fn after_action,
                                                          {query, {:ok, results, notifications}} ->
       case after_action.(query, results) do
         {:ok, results} ->
-          {:cont, {query, {:ok, results, notifications}}}
+          {:cont, {set_phase(query), {:ok, results, notifications}}}
```

Comment:
I don't think we want to reset the phase here, since its still in the `after_action`.